### PR TITLE
Add schema generation benchmarks for models with custom serializers

### LINKED ISSUE #9711: Expand breadth of benchmarks
We have a robust benchmark test suite in `pydantic-core`, but are lacking similar coverage in `pydantic`.

Specifically, I'm interested in expanding this test suite in relation to our core schema build times, as we're looking to improve performance there in the short term. Secondarily, I'm interested in better benchmarking for our json schema generation. Benchmarks for validation and serialization are great, but we've already got pretty good coverage in `pydantic-core` for those, so I'm less concerned about such benchmarks.

This is a great first issue - you can help write tests that test a variety of use cases for `pydantic`, without having to delve into the internals of `pydantic` code to fix bugs or add new features.

Here are some cases for which I think we should expand our suite of benchmarks (assume these are for core schema generation, but you can write related tests for any of the above mentioned categories). As in, you could read the first bullet as "we need to write some tests showcasing how long schema building takes for models with custom validators".

- [x] custom validators (field and model, also before, wrap, after, and plain)
- [x] tagged and untagged unions
- [x] stdlib types
- [x] second class Pydantic types like `SecretStr` and `PositiveInt`
- [x] custom serializers (field and model, also before, wrap, after, and plain)
- [x] recursive model definitions -- need much more robust testing when defs / refs are involved
- [ ] third party, custom types

I wouldn't say these are necessary to close this issue, but one thing for more involved contributors to look at would be benchmarking specific parts of our core schema generation process in `pydantic`, such as:

- [ ] annotation application
- [ ] discriminator application
- [ ] schema cleaning (including defs / refs simplification)
- [ ] schema walking
- [ ] field acquisition / setup in `ModelMetaClass.__new__`

### PR DESCRIPTION
**Pull request by AdolfoVillalobos**
## Change Summary

This PR expands the benchmark suite by adding "schema generation" benchmarks in the following cases:

1. Models with custom field serializers (decorator and annotated)
2. Models with custom model serializers

I am open to suggestions.

## Related issue number

Contributes to #9711 

## Checklist

* [ ] The pull request title is a good summary of the changes - it will be used in the changelog
* [ ] Unit tests for the changes exist
* [ ] Tests pass on CI
* [ ] Documentation reflects the changes where applicable
* [ ] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

### CODE REVIEW

**Comment by AdolfoVillalobos** [comment]
@sydney-runkle I created benchmarks for the custom serializer case in schema generation and incorporated some of the cool parametric features that you added to your refactor. Very cool stuff!

Let me know if there is anything else you want added to this PR or if there is some issue I'm not seeing.

File: tests/benchmarks/test_model_schema_generation.py [hunk 1]
```diff
@@ -262,3 +267,54 @@ def validate_model_wrap(cls, values: Any, handler: ModelWrapValidatorHandler[Sel
...
+def test_custom_field_serializer(benchmark, mode) -> None:
+    def schema_gen() -> None:
+        class ModelWithFieldSerializer(BaseModel):
+            field1: int
+
+            @field_serializer('field1', mode=mode)
+            def serialize_field(cls, v: int) -> str:
+                return str(v)
+
+    benchmark(schema_gen)
```
Comment (sydney-runkle) [inline tests/benchmarks/test_model_schema_generation.py:283]: I think we need to separate the tests for `plain` and `wrap` serializers - they have different signatures (ex - see the different signatures for the corresponding validators). This looks like a great test for the `plain` serializer, though.

Comment (sydney-runkle) [reply tests/benchmarks/test_model_schema_generation.py:283]: (Just like you did for the annotated ones).

**Review by sydney-runkle — CHANGES_REQUESTED** [review]
This looks great - just one follow up request.

@AdolfoVillalobos - truly great work here - so excited to see our benchmark suite systematically expanding. Nice work doing this in small PRs as well - makes things incredibly easy to review 🚀

File: tests/benchmarks/test_model_schema_generation.py (hunk 1 above)
Comment (AdolfoVillalobos) [reply tests/benchmarks/test_model_schema_generation.py]: You are right! I forgot about the signature difference. I believe it should be correct now.

**Review by sydney-runkle — APPROVED** [review]

**Comment by sydney-runkle** [comment]
Great work @AdolfoVillalobos, thanks for the consistent progress here!

**Pull request merged**